node/src/blockchain/blockchain_interface/blockchain_interface_web3/lower_level_interface_web3.rs

```diff
@@ -1,395 +1,762 @@
+    use crate::blockchain::blockchain_interface::blockchain_interface_web3::lower_level_interface_web3::TransactionReceiptResult;
+    use crate::blockchain::blockchain_interface::data_structures::errors::BlockchainError::QueryFailed;
+    use crate::blockchain::test_utils::{make_blockchain_interface_web3, make_tx_hash, ReceiptResponseBuilder};
+    use crate::test_utils::{assert_string_contains, make_wallet};
 
     #[test]
-    fn low_interface_web3_transaction_fee_balance_works() {
+    fn get_transaction_fee_balance_works() {
         let port = find_free_port();
-        let test_server = TestServer::start(
-            port,
-            vec![br#"{"jsonrpc":"2.0","id":0,"result":"0xDEADBEEF"}"#.to_vec()],
-        );
-        let (_event_loop_handle, transport) = Http::with_max_parallel(
-            &format!("http://{}:{}", &Ipv4Addr::LOCALHOST.to_string(), port),
-            REQUESTS_IN_PARALLEL,
-        )
-        .unwrap();
-        let chain = TEST_DEFAULT_CHAIN;
-        let subject = make_subject(transport, chain);
+        let _blockchain_client_server = MBCSBuilder::new(port)
+            .response("0x23".to_string(), 1)
+            .start();
+        let wallet = &Wallet::from_str("0x3f69f9efd4f2592fd70be8c32ecd9dce71c472fc").unwrap();
+        let subject = make_blockchain_interface_web3(Some(port));
+
+        let result = subject
+            .lower_interface()
+            .get_transaction_fee_balance(wallet.address())
+            .wait();
+
+        assert_eq!(result, Ok(35.into()));
+    }
+
+    #[test]
+    fn get_transaction_fee_balance_returns_an_error_for_unintelligible_response_to_requesting_eth_balance(
+    ) {
+        let port = find_free_port();
+        let _blockchain_client_server = MBCSBuilder::new(port)
+            .response("0xFFFQ".to_string(), 0)
+            .start();
+        let subject = make_blockchain_interface_web3(Some(port));
 
         let result = subject
+            .lower_interface()
             .get_transaction_fee_balance(
-                &Wallet::from_str("0x3f69f9efd4f2592fd70be8c32ecd9dce71c472fc").unwrap(),
+                Wallet::from_str("0x3f69f9efd4f2592fd70be8c32ecd9dce71c472fc")
+                    .unwrap()
+                    .address(),
             )
-            .unwrap();
+            .wait();
+
+        match result {
+            Err(BlockchainError::QueryFailed(msg)) if msg.contains("invalid hex character: Q") => {
+                ()
+            }
+            x => panic!("Expected complaint about hex character, but got {:?}", x),
+        };
+    }
+
+    #[test]
+    fn get_gas_price_works() {
+        let port = find_free_port();
+        let _blockchain_client_server = MBCSBuilder::new(port)
+            .response("0x01".to_string(), 1)
+            .start();
+        let subject = make_blockchain_interface_web3(Some(port));
+
+        let result = subject.lower_interface().get_gas_price().wait().unwrap();
+
+        assert_eq!(result, 1.into());
```

Comment:
The most important part, the assertion checking what json was sent out, is missing. Pull it out as a record from the mock server and find a way to assert on it. Especially important piece of string is the opcode of the request produced.